walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,129 @@
+    
+    long duration = request.getDuration();
+
+    // Meetings can't be longer than a day
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return options;
+    } else {
+      options.add(TimeRange.WHOLE_DAY);
+    }
+
+    ArrayList<TimeRange> allUnavailableBlocks;
+    allAttendees.addAll(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+
+    Collection<String> optAttendees = request.getOptionalAttendees();
+    Collection<String> reqAttendees = request.getAttendees();
+    
+    // Do not need to check if optional attendees are compatible with meeting request
+    if (reqAttendees.size() == 0 || optAttendees.size() == 0){
+      allUnavailableBlocks = unavailableTime(events, allAttendees, duration);
+      findFreeTime(options, allUnavailableBlocks, duration);
+
+    } else {
+      ArrayList<TimeRange> reqUnavailableBlocks = unavailableTime(events, reqAttendees, duration);
+      allUnavailableBlocks = unavailableTime(events, allAttendees, duration);
+      
+      // Preserve options in case cannot accommodate additional opt attendees
+      ArrayList<TimeRange> optionsTemp = findFreeTime(new ArrayList<>(options), allUnavailableBlocks, duration);
+      if (optionsTemp.size() == 0) { // opt attendees not compatible, focus on req attendees
+        optionsTemp = findFreeTime(new ArrayList<>(options), reqUnavailableBlocks, duration);
+      }
+      options = optionsTemp;
+    }
+    return options;
+  }
+
+  /**
+   * Find all possible time ranges for meeting 
+   */
+  private ArrayList<TimeRange> findFreeTime(ArrayList<TimeRange> options, ArrayList<TimeRange> blockedTime, long duration) {
+
+    for (TimeRange block : blockedTime) {
+      int busyStart = block.start();
+      int busyEnd = block.end();
+
+      for (TimeRange freeTime : new ArrayList<>(options)) {
+
+        // Conflict in ranges
+        if (block.overlaps(freeTime)) {
+          int freeStart = freeTime.start();
+          int freeEnd = freeTime.end();
+
+          // Free time extends beyond busy time
+          if (!block.contains(freeTime)) {
+            // Check for free time before event starts
+            if ((freeStart + duration) <= busyStart) {
+              options.add(TimeRange.fromStartEnd(freeStart, busyStart, false));
+            }
+            // Check for free time after event ends
+            if ((busyEnd + duration) <= freeEnd) {
+              options.add(TimeRange.fromStartEnd(busyEnd, freeEnd, false));
+            }
+          }
+          options.remove(freeTime);
+        }
+      } 
+    }
+    return options;
+  }
+
+  /**
+   * Block off all unavailable time
+   */
+  private ArrayList<TimeRange> unavailableTime(Collection<Event> allEvents, Collection<String> attendees, long duration) {
+    ArrayList<TimeRange> blockedTimeRanges = findAttendeesBlockedTimeRanges(allEvents, attendees);
+    return mergeTimeRanges(blockedTimeRanges, duration);
```

Comment:
Completely non-actionable, but fun for future algorithmic experiments: reducing input size (by merging blocks in this case) and pre-sorting in general can be a nice optimization, but in this case it will actually slow the code down. Merging, due to the sort required, takes O(n*log(n)), where n = number of busy meetings. Without merging, your algorithm still works perfectly, but it's O(n * m), where `m` is the number of free slots. Observe, however, that `m` trends downward to 0 very quickly as input grows because you remove too-small free slots as soon as you see them.

```
Without calling merge (just straight returning `blockedTimeRanges`):
With 1 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 3
With 10 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 8
With 100 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 10
With 1000 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 30
With 10000 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 114



With calling merge (code as is):
With 1 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 4
With 10 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 7
With 100 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 12
With 1000 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 103
With 10000 busy meetings, finding a meeting slot 100 times takes (in milliseconds): 3316
```